### PR TITLE
[Silabs] [EFR32] Adds RS9116 BLE enabled for all applications 

### PR DESCRIPTION
**Pull request by bhmanda-silabs**
Problem
What is being fixed? Examples:
Earlier all the applications use EFR32 BLE with RS9116 boards. 

Change Overview
In this PR, we enabled RS9116 BLE for all the applications. 

Testing
Tested manually by using MG12+RS9116
Tested manually by using MG24+RS9116

### CODE REVIEW

File: examples/lighting-app/silabs/efr32/BUILD.gn [hunk 1]
```diff
@@ -56,6 +56,12 @@ efr32_sdk("sdk") {
     "${examples_common_plat_dir}",
   ]
 
+  if (use_rs911x) {
```
Comment (jepenven-silabs) [inline examples/lighting-app/silabs/efr32/BUILD.gn:59]: This shouldn't go there. @jmartinez-silabs did a cleanup of all of the GN files. It should go in the comment platform.

File: examples/thermostat/silabs/efr32/BUILD.gn [hunk 2]
```diff
@@ -60,6 +60,12 @@ efr32_sdk("sdk") {
     "${examples_common_plat_dir}",
   ]
 
+  if (use_rs911x) {
```
Comment (jepenven-silabs) [inline examples/thermostat/silabs/efr32/BUILD.gn:63]: Shouldn't be there.

File: examples/window-app/silabs/efr32/BUILD.gn [hunk 3]
```diff
@@ -50,6 +50,12 @@ efr32_sdk("sdk") {
     "${examples_common_plat_dir}",
   ]
 
+  if (use_rs911x) {
```
Comment (jepenven-silabs) [inline examples/window-app/silabs/efr32/BUILD.gn:53]: Shouldn't be there.

**Review by jepenven-silabs — CHANGES_REQUESTED** [review]

File: src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp [hunk 4]
```diff
@@ -0,0 +1,1008 @@
...
+        case RSI_BLE_CONN_EVENT: {
+            rsi_ble_app_clear_event(RSI_BLE_CONN_EVENT);
+            chip::DeviceLayer::Internal::BLEMgrImpl().HandleConnectEvent();
+            WFX_RSI_LOG("%s Module got connected", __func__);
+        }
+        break;
+        case RSI_BLE_DISCONN_EVENT: {
+            // event invokes when disconnection was completed
+            WFX_RSI_LOG("%s Module got Disconnected", __func__);
+            chip::DeviceLayer::Internal::BLEMgrImpl().HandleConnectionCloseEvent(event_msg.reason);
```
Comment (jmartinez-silabs) [inline src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp:131]: using the namespace allow you to reduce the function call lenght
`BLEMgrImpl().HandleConnectionCloseEvent`

File: src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp [hunk 5]
```diff
@@ -0,0 +1,1008 @@
...
+void BLEManagerImpl::NotifyChipConnectionClosed(BLE_CONNECTION_OBJECT conId)
+{
+    // Nothing to do
+}
+
+CHIP_ERROR BLEManagerImpl::MapBLEError(int bleErr)
+{
+    switch (bleErr)
+    {
+    case SL_STATUS_OK:
```
Comment (jmartinez-silabs) [inline src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp:477]: does the rs911x return SL_STATUS codes? if not then this whole case as not purpose.

File: src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp [hunk 6]
```diff
@@ -0,0 +1,1008 @@
...
+{
+    CHIP_ERROR err;
+    int32_t status = 0;
+
+    ChipLogProgress(DeviceLayer, "StartAdvertising start");
+
+    // If already advertising, stop it, before changing values
+    if (mFlags.Has(Flags::kAdvertising))
+    {
+        // sl_bt_advertiser_stop(advertising_set_handle);
```
Comment (jmartinez-silabs) [inline src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp:617]: with this commented this if else should be change to just a 
```
if (!mFlags.Has(Flags::kAdvertising))
{
    ChipLogDetail(DeviceLayer, "Start BLE advertissement");
}
```

File: src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp [hunk 7]
```diff
@@ -0,0 +1,1008 @@
...
+    mFlags.Set(Flags::kEFRBLEStackInitialized);
+}
+
+void BLEManagerImpl::HandleConnectEvent(void)
+{
+    ChipLogProgress(DeviceLayer, "Connect Event for handle : %d", event_msg.connectionHandle);
+    AddConnection(event_msg.connectionHandle, event_msg.bondingHandle);
+}
+
+// TODO:: Implementation need to be done.
```
Comment (jmartinez-silabs) [inline src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp:705]: is this comment still true ? it seems implemented

File: src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp [hunk 8]
```diff
@@ -0,0 +1,1008 @@
...
+        WFX_RSI_LOG("Inside HandleTXCharCCCDWrite ");
+        HandleTXCharCCCDWrite(&evt);
+    }
+    else
+    {
+        HandleRXCharWrite(&evt);
+    }
+}
+
+//TODO:: Need to implement this
```
Comment (jmartinez-silabs) [inline src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp:761]: same here

File: src/platform/silabs/efr32/rs911x/wfx_sl_ble_init.c [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+ * @section description
+ * This function is used at application to create new service.
+ */
+
+void rsi_ble_add_char_val_att(void * serv_handler, uint16_t handle, uuid_t att_type_uuid, uint8_t val_prop, uint8_t * data,
+                                     uint8_t data_len, uint8_t auth_read)
+{
+    rsi_ble_req_add_att_t new_att = { 0 };
+
+    memset(&new_att, 0, sizeof(rsi_ble_req_add_att_t));
```
Comment (jmartinez-silabs) [inline src/platform/silabs/efr32/rs911x/wfx_sl_ble_init.c:317]: no need for the memset 0 you init the struct to = { 0 } just before.

File: third_party/silabs/efr32_sdk.gni [hunk 10]
```diff
@@ -31,6 +33,8 @@ declare_args() {
 
   kvs_max_entries = 75
 
+  chip_enable_ble_rs911x = false
```
Comment (jmartinez-silabs) [inline third_party/silabs/efr32_sdk.gni:36]: move this to the declare_args in `silabs_board.gni` as it was done for other board/hardware specific argument

File: third_party/silabs/efr32_sdk.gni [hunk 11]
```diff
@@ -191,6 +204,16 @@ template("efr32_sdk") {
       #"__STACK_SIZE=0",
     ]
 
+    if (chip_enable_ble_rs911x) {
```
Comment (jmartinez-silabs) [inline third_party/silabs/efr32_sdk.gni:207]: Those two define have the same if condition. Don't split it in two if. regroup them.

Also why do you need `RSI_BLE_ENABLE`=1 and `RS91X_BLE_ENABLE`=1 why do you need two defines that are effectively the same

File: third_party/silabs/efr32_sdk.gni [hunk 12]
```diff
@@ -156,6 +160,15 @@ template("efr32_sdk") {
       ]
     }
 
+    if (chip_enable_ble_rs911x) {
+      _include_dirs += [
+        "${chip_root}/examples/platform/silabs/efr32/rs911x",
```
Comment (jmartinez-silabs) [inline third_party/silabs/efr32_sdk.gni:165]: include_dir of `${chip_root}/examples/platform/silabs` and ${chip_root}/src/platform/silabs/efr32/rs911x should probably not be in here. only `sdk_support_root` and `efr32_sdk_root` should be linked here.

Why do you need them?
`${chip_root}/examples/platform/silabs` is already part of the source_sets in `${chip_root}/examples/platform/silabs/BUILD.gn`

**Review by jmartinez-silabs — CHANGES_REQUESTED** [review]

File: src/platform/silabs/efr32/rs911x/wfx_sl_ble_init.c [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+ * @return     status (uint32_t) 0 for success.
+ * @section description
+ * This function is used at application to create new service.
+ */
+
+uint32_t rsi_ble_add_matter_service(void)
+{
+    uuid_t custom_service    = { RSI_BLE_MATTER_CUSTOM_SERVICE_UUID };
+    custom_service.size      = RSI_BLE_MATTER_CUSTOM_SERVICE_SIZE;
+    custom_service.val.val16 = RSI_BLE_MATTER_CUSTOM_SERVICE_VALUE_16;
```
Comment (jmeg-sfy) [inline src/platform/silabs/efr32/rs911x/wfx_sl_ble_init.c:377]: What is this service exactly ?

**Review by jmeg-sfy — APPROVED** [review]

**Review by jmeg-sfy — COMMENTED** [review]
I think you should follow junior guidance

**Review by jmeg-sfy — CHANGES_REQUESTED** [review]
Please follow comments

File: examples/lighting-app/silabs/efr32/BUILD.gn (hunk 1 above)
Comment (shgutte) [reply examples/lighting-app/silabs/efr32/BUILD.gn:59]: Okay @jepenven-silabs we don't know about the @jmartinez-silabs PR we moved this to efr32_sdk.gni

File: examples/lock-app/silabs/efr32/BUILD.gn [hunk 14]
```diff
@@ -56,6 +56,12 @@ efr32_sdk("sdk") {
     "${examples_common_plat_dir}",
   ]
 
+  if (use_rs911x) {
```
[the remark this answers is not in the document]
Comment (shgutte) [reply examples/lock-app/silabs/efr32/BUILD.gn:59]: Done

File: third_party/silabs/efr32_sdk.gni (hunk 11 above)
Comment (shgutte) [reply third_party/silabs/efr32_sdk.gni:207]: we will combine both "if" conditions.
RSI_BLE_ENABLE is used in the SAPI code of wiseconnect to enable the BLE on RS9116. The macro RS91X_BLE_ENABLE is used on the matter side examples/platform code separately because if someone changed the macro on SAPI we should not need to do matter side code changes only need to change the macro definition.

File: src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp (hunk 7 above)
Comment (bhmanda-silabs) [reply src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp:706]: @jmartinez-silabs yeah it is implemented so removed the comment.

File: src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp (hunk 8 above)
Comment (bhmanda-silabs) [reply src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp:762]: removed the TODO line.

File: src/platform/silabs/efr32/BUILD.gn [hunk 15]
```diff
@@ -66,12 +68,23 @@ static_library("efr32") {
...
 
+  if (chip_enable_ble_rs911x) {
+    sources += [
+        "rs911x/BLEManagerImpl.cpp",
+        "rs911x/wfx_sl_ble_init.h",
+        "rs911x/wfx_sl_ble_init.c",
+        "rs911x/rsi_ble_config.h",
+      ]
+    } else {
+    sources += [ "BLEManagerImpl.cpp" ]
```
[the remark this answers is not in the document]
Comment (bhmanda-silabs) [reply src/platform/silabs/efr32/BUILD.gn:82]: Because in the RS9116 before initializing the BLE we should initialize the Driver, WiSeConnect and radio that's why there is implementation difference.

File: src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp (hunk 5 above)
Comment (bhmanda-silabs) [reply src/platform/silabs/efr32/rs911x/BLEManagerImpl.cpp:478]: We already have jira task for this , we will implement in the coming PR. 
[](https://jira.silabs.com/browse/MATTER-1406 )

File: src/platform/silabs/efr32/BLEManagerImpl.h [hunk 16]
```diff
@@ -0,0 +1,232 @@
+/*
```
[the remark this answers is not in the document]
Comment (bhmanda-silabs) [reply src/platform/silabs/efr32/BLEManagerImpl.h:1]: We did not know that, in folder structure clean up this header file has been moved to **src/platform/silabs**. Now we removed this file.

File: third_party/silabs/efr32_sdk.gni (hunk 11 above)
Comment (jmartinez-silabs) [reply third_party/silabs/efr32_sdk.gni:205]: I would advocate to actually try to reuse the SAPI defines. Even if they change in SAPI, you aren't updating the sdk(wiseconnect) that often. If you do update it, you would go over test and validation I assume and can catch the change and address it in your update sdk pr. 


I am saying this because we are starting to have so many defines and args that it is hard to follow. Same thing for our thread platform by the way. If we can share defines from gsdk to our matter platform code we should

File: examples/platform/silabs/efr32/rs911x/rsi_if.c [hunk 17]
```diff
@@ -73,6 +73,10 @@ bool is_wifi_disconnection_event = false;
 /* Declare a variable to hold connection time intervals */
 uint32_t retryInterval = WLAN_MIN_RETRY_TIMER_MS;
 
+#if (RS91X_BLE_ENABLE)
+extern rsi_semaphore_handle_t sl_ble_init_sem;
```
[the remark this answers is not in the document]
Comment (bhmanda-silabs) [reply examples/platform/silabs/efr32/rs911x/rsi_if.c]: @jepenven-silabs we have changed the name of the semaphore.

File: src/platform/silabs/efr32/rs911x/wfx_sl_ble_init.c (hunk 13 above)
Comment (bhmanda-silabs) [reply src/platform/silabs/efr32/rs911x/wfx_sl_ble_init.c:375]: @jmeg-sfy This custom GATT service has been created for data transfer. RS911x module maintains the GATT service on the host , so created in the application context.

File: examples/lock-app/silabs/efr32/BUILD.gn [hunk 18]
```diff
@@ -61,6 +62,11 @@ efr32_sdk("sdk") {
     include_dirs += [ "${examples_plat_dir}/wf200" ]
   }
 
+  if (chip_enable_ble_rs911x) {
```
Comment (jepenven-silabs) [inline examples/lock-app/silabs/efr32/BUILD.gn:65]: This goes against @jmartinez-silabs cleanup no ?

File: examples/thermostat/silabs/efr32/BUILD.gn [hunk 19]
```diff
@@ -65,6 +66,11 @@ efr32_sdk("sdk") {
     include_dirs += [ "${examples_plat_dir}/wf200" ]
   }
 
+  if (chip_enable_ble_rs911x) {
```
Comment (jepenven-silabs) [inline examples/thermostat/silabs/efr32/BUILD.gn:69]: same comment

**Review by jepenven-silabs — CHANGES_REQUESTED** [review]
A lot of duplication is created with this PR in High Level code. This code should make abstraction of the hardware used by default. The fact that a duplication seems to be needed highlight that there might be issues with the level of abstraction or a bug in the odering of the inits. 
in either cases, duplicating source code isn't the way to go. Both hardware should be able to use the same High Level classes.

File: examples/lock-app/silabs/efr32/BUILD.gn (hunk 18 above)
Comment (jmartinez-silabs) [reply examples/lock-app/silabs/efr32/BUILD.gn:65]: Best solution currently. Will have to refactore and move some files in they created in that include dir

File: third_party/silabs/efr32_sdk.gni (hunk 12 above)
Comment (jmartinez-silabs) [reply third_party/silabs/efr32_sdk.gni:163]: You added it back to the efr32_sdk() in the examples build.gn. So it should not be repeated here

**Review by jepenven-silabs — APPROVED** [review]
Approved considering this will be refactor soon.

**Review by jmartinez-silabs — APPROVED** [review]

**Review by jmeg-sfy — APPROVED** [review]
Approved since i heard refactoring are in the pipe thereafter
